linkerd/core/src/main/scala/io/buoyant/linkerd/RequestEvaluator.scala

```diff
@@ -0,0 +1,165 @@
+import io.buoyant.namer.DelegateTree._
+import io.buoyant.namer.{DelegateTree, Delegator}
+import io.buoyant.router.RoutingFactory
+import io.buoyant.router.RoutingFactory.BaseDtab
+import io.buoyant.router.context.{DstBoundCtx, DstPathCtx}
+
+case class EvaluatedRequest(
+  identification: String,
+  selectedAddress: String,
+  addresses: Option[Set[String]],
+  dtabResolution: List[String]
+) {
+  override def toString() = {
+    s"""
+                                 |identification: $identification
+                                 |selectedAddress: $selectedAddress
+                                 |addresses: ${addresses.getOrElse(Set.empty).mkString(",")}
+                                 |Dtab Resolution:
+                                 |${dtabResolution.mkString("\n")}
+    """.stripMargin
+  }
+}
+
+class RequestEvaluator(
+  endpoint: EndpointAddr,
+  namer: DstBindingFactory.Namer,
+  dtab: BaseDtab
+) extends SimpleFilter[Request, Response] {
+
+  private val EvaluatorHeaderName = "l5d-req-evaluate"
+  private val JsonCharSet = ";charset=UTF-8"
+
+  override def apply(
+    req: Request,
+    svc: Service[Request, Response]
+  ): Future[Response] = {
+
+    req.headerMap.get(EvaluatorHeaderName) match {
+      case None => svc(req)
+      case Some(_) =>
+
+        val identificationPath = DstPathCtx.current match {
+          case Some(dstPath) => dstPath.path
+          case None => Path.empty
+        }
+
+        val selectedEndpoint = endpoint.addr match {
+          case inetAddr: Address.Inet => inetAddr.addr.toString
+          case _ => ""
+        }
+
+        val lbSet = DstBoundCtx.current match {
+          case None => Addr.Neg
+          case Some(addrSet) => addrSet.name.addr.sample()
+        }
+
+        val addresses = lbSet match {
+          case Addr.Neg | Addr.Failed(_) | Addr.Pending => None
+          case Addr.Bound(a, _) => Some(
+            a.map { address: Address =>
+              address match {
+                case inetAddr: Address.Inet => inetAddr.addr.toString
+                case _ => ""
+              }
+            }
+          )
+        }
+
+        val dtreeF = namer.interpreter match {
+          case delegator: Delegator => delegator.delegate(dtab.dtab(), identificationPath)
+            .map(Some(_))
+          case _ => Future.None
+        }
+
+        dtreeF.map { dtree =>
+          val resp = Response()
+          val tree = RequestEvaluator.formatDTree(dtree, req.contentType, List.empty)
+          val evaluatedRequest = EvaluatedRequest(
+            Try(tree.last) match {
+              case Throw(_) => "Unknown identification"
+              case Return(id) => id
+            },
+            selectedEndpoint,
+            addresses,
+            tree
+          )
+
+          resp.contentType = req.contentType.getOrElse(MediaType.PlainText)
+          resp.contentString = RequestEvaluator
+            .writeContentString(resp.contentType, evaluatedRequest)
```

Comment:
I was never super clear on this... should this key off the request's `accept` header instead?

Reply:
You're right. I mistook the `Content-Type` header for the `Accept` header.